vpn/src/main/java/com/duckduckgo/mobile/android/vpn/service/TrackerBlockingVpnService.kt

```diff
@@ -307,6 +290,70 @@ class TrackerBlockingVpnService : VpnService(), CoroutineScope by MainScope(), N
         }
     }
 
+    private fun getDns(): Set<InetAddress> {
+        // private extension function, this is purposely here to limit visibility
+        fun Set<InetAddress>.containsIpv4(): Boolean {
+            forEach {
+                if (it is Inet4Address) return true
+                return false
+            }
+            return false
+        }
+
+        val dns = mutableSetOf<InetAddress>()
+
+        // Default system DNS
+        if (appTpFeatureConfig.isEnabled(AppTpSetting.SetActiveNetworkDns)) {
+            kotlin.runCatching {
+                applicationContext.getSystemActiveNetworkDefaultDns()
+                    .map { InetAddress.getByName(it) }
+            }.getOrNull()?.run { dns.addAll(this) }
+        }
+
+        // Android Private DNS (added by the user)
+        if (appTpFeatureConfig.isEnabled(AppTpSetting.PrivateDnsSupport)) {
+            vpnPreferences.privateDns?.let { privateDnsName ->
+                runCatching {
+                    InetAddress.getAllByName(privateDnsName)
+                }.getOrNull()?.run { dns.addAll(this) }
+            }
+        }
+
+        if (appTpFeatureConfig.isEnabled(AppTpSetting.AlwaysSetDNS)) {
+            // Always add DNS
+            if (dns.isEmpty()) {
```

Comment:
does this mean AlwaysSetDNS and SetActiveNetworkDns are not compatible? `dns` won't be empty at this point if `applicationContext.getSystemActiveNetworkDefaultDns` is not empty.

Reply:
`AlwaysSetDNS` and `SetActiveNetworkDns` are independent, we could have `SetActiveNetworkDns` disabled and then the dns list could be empty.
`AlwaysSetDNS` is only meant for internal use only, not production, in case we need to check certain type of breakage